client/asset/btc/btc_test.go

```diff
@@ -2281,3 +2311,116 @@ func TestSyncStatus(t *testing.T) {
 		t.Fatalf("progress out of range. Expected 0.5, got %.2f", progress)
 	}
 }
+
+func TestPreSwap(t *testing.T) {
+	wallet, node, shutdown, err := tNewWallet(false)
+	defer shutdown()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// See math from TestFundEdges. 10 lots with max fee rate of 34 sats/vbyte.
+
+	swapVal := uint64(1e7)
+	lots := swapVal / tBTC.LotSize // 10 lots
+
+	const swapSize = 225
+	const totalBytes = 2250
+	const bestCaseBytes = 513
+	var estFeeRate = optimalFeeRate + 1 // +1 added in feeRate
+
+	backingFees := uint64(totalBytes) * tBTC.MaxFeeRate // total_bytes * fee_rate
+
+	minReq := swapVal + backingFees
+
+	p2pkhUnspent := &ListUnspentResult{
+		TxID:          tTxID,
+		Address:       tP2PKHAddr,
+		Confirmations: 5,
+		ScriptPubKey:  tP2PKH,
+		Spendable:     true,
+		Solvable:      true,
+		Safe:          true,
+	}
+	unspents := []*ListUnspentResult{p2pkhUnspent}
+
+	setFunds := func(v uint64) {
+		p2pkhUnspent.Amount = float64(v) / 1e8
+		node.rawRes[methodListUnspent] = mustMarshal(t, unspents)
+	}
+
+	form := &asset.PreSwapForm{
+		LotSize:     tBTC.LotSize,
+		Lots:        lots,
+		AssetConfig: tBTC,
+		Immediate:   false,
+	}
+
+	setFunds(minReq)
+
+	// Initial success.
+	preSwap, err := wallet.PreSwap(form)
+	if err != nil {
+		t.Fatalf("PreSwap error: %v", err)
+	}
+
+	maxFees := totalBytes * tBTC.MaxFeeRate
+	estHighFees := totalBytes * estFeeRate
+	estLowFees := bestCaseBytes * estFeeRate
+	checkSwapEstimate(t, preSwap.Estimate, lots, swapVal, maxFees, estHighFees, estLowFees, minReq)
+
+	// Too little funding is an error.
+	setFunds(minReq - 1)
+	_, err = wallet.PreSwap(form)
+	if err == nil {
+		t.Fatalf("no PreSwap error for not enough funds: %v", err)
+	}
+	setFunds(minReq)
+
+	node.estFeeErr = tErr
+	_, err = wallet.PreSwap(form)
+	if err == nil {
+		t.Fatalf("no PreSwap error for not enough funds: %v", err)
+	}
+	node.estFeeErr = nil
```

Comment:
Minor conflict with the `if c.estFeeErr != nil { return nil, c.estFeeErr }` needing to be in the `RawRequest` stub now.  Also looks like the `Fatalf` message here is copy-pasta'd from the previous test case.